Rank Heatmaps

nbokulich opened the pull request:
partially fixes #53 

This generates rank heatmaps, and registers the function as a QIIME 2 visualizer.

mortonjt commented on rhapsody/heatmap.py:
```diff
@@ -0,0 +1,157 @@
+                  x_labels=False, y_labels=False, level=3, threshold=3):
+    '''
+    Generate clustermap of microbe X metabolite conditional probabilities.
+
```
```suggestion
Parameters
----------
```

mortonjt commented on rhapsody/heatmap.py:
```diff
@@ -0,0 +1,157 @@
+        Minimum absolute value of conditional probabilities to plot.
+        metabolites/taxa that do not have at least one CP above the
+        threshold will be removed.
+
```
Returns
--------
sns.clustermap

mortonjt commented on rhapsody/heatmap.py:
```diff
@@ -0,0 +1,157 @@
+    '''
+    # filter microbes
+    abs_vals = ranks.abs()
+    ranks = ranks[abs_vals.max(axis=1) > threshold]
```
This is one of the more challenging parts for the UI.

First, definitely don't want to take the max of the absolutes - we don't care about the strongly negative conditional probabilities (those are typically not informative). 

Since the user is already passing in metadata, maybe we could just have the ranks just match the metadata and only consider the intersection?  The filtering criteria can get very complicated otherwise (i.e. for instance, the threshold is expected not to be reliable compared to the ranks.  Also the negative conditional log probabilities won't be informative, so we won't want to take the absolute values of that).

mortonjt commented on rhapsody/heatmap.py:
```diff
@@ -0,0 +1,157 @@
+    hotmap = sns.clustermap(ranks, cmap=color_palette, center=0,
+                            col_colors=col_colors, row_colors=row_colors,
+                            figsize=(12, 12), method=method, metric=metric,
+                            cbar_kws={'label': 'Conditional\nProbability'})
```
`cbar_kws={'label': 'Conditional\nProbability'})` -> `cbar_kws={'label': 'Log Conditional\nProbability'})`

nbokulich replied:
I am adding this in the changes I will push up soon, but I am conflicted. Will the `Conditionals` always contain log conditional probabilities? Since currently the answer is yes, I have made this change, but we may want to consider if you foresee other types of data being contained in the `Conditionals`.

mortonjt commented on rhapsody/heatmap.py:
```diff
@@ -0,0 +1,157 @@
+    '''
+    Transform feature or sample metadata into color vector for annotating
+    margin of clustermap.
+
```
```suggestion
Parameters
----------
```

mortonjt commented on rhapsody/heatmap.py:
```diff
@@ -0,0 +1,157 @@
+    margin_palette: str
+        Name of color palette to use for annotating metadata
+        along margin(s) of clustermap.
+
```
```suggestion
Returns
-------
```

mortonjt commented on rhapsody/heatmap.py:
```diff
@@ -0,0 +1,157 @@
+    '''
+    taxonomy_series: pd.Series of semicolon-delimited taxonomy strings
+    level: int
+        taxonomic level for annotating clustermap.
```
```suggestion
        taxonomic level for annotating clustermap.
 Returns
 -------
```

mortonjt commented on rhapsody/heatmap.py:
```diff
@@ -0,0 +1,157 @@
+        metabolites/taxa that do not have at least one CP above the
+        threshold will be removed.
+
+    Returns clustermap
```
```suggestion
    Returns
    -------
    sns.clustermap
```

mortonjt commented on rhapsody/heatmap.py:
```diff
@@ -0,0 +1,157 @@
+    abs_vals = ranks.abs()
+    ranks = ranks[abs_vals.max(axis=1) > threshold]
+    # filter metabolites
+    ranks = ranks[ranks.columns[(abs_vals > threshold).any()]]
```
Same here regarding the thresholding according to ranks - I think we can defer this to the metadata (so we can leave this out for now).

mortonjt commented on rhapsody/q2/plugin_setup.py:
```diff
@@ -86,6 +87,46 @@
+        'threshold': 'Minimum absolute value of conditional probabilities to '
+                     'plot. Metabolites/taxa that do not have at least one '
+                     'conditional probability above the threshold will be '
+                     'removed.'
```
Let's aim to remove this parameter for now

nbokulich replied:
👍

mortonjt commented on rhapsody/q2/tests/test_visualizers.py:
```diff
@@ -0,0 +1,49 @@
+        rhapsody.actions.heatmap(self.ranks, self.taxa, None, level=-1)
+
+    def test_heatmap_plot_axis_labels(self):
+        rhapsody.actions.heatmap(self.ranks, x_labels=True, y_labels=True)
```
These are really awesome tests.  Thank you!

mortonjt commented:
I think it should be ok to only have log conditional probabilities. This
type is specific to rhapsody for more, so hopefully that shouldn't be a
problem

On Wed, Jul 31, 2019, 9:09 AM Nicholas Bokulich <notifications@github.com>
wrote:

> *@nbokulich* commented on this pull request.
> ------------------------------
>
> In rhapsody/heatmap.py
> <https://github.com/biocore/rhapsody/pull/69#discussion_r309209841>:
>
> > +        row_colors = None
> +
> +    # subset metabolite metadata based on rows/columns
> +    if metabolite_metadata is not None:
> +        metabolite_metadata = metabolite_metadata.reindex(ranks.columns)
> +        # map metadata categories to column colors
> +        col_colors, col_class_colors = _parse_heatmap_metadata_annotations(
> +            metabolite_metadata, margin_palette)
> +    else:
> +        col_colors = None
> +
> +    # Generate heatmap
> +    hotmap = sns.clustermap(ranks, cmap=color_palette, center=0,
> +                            col_colors=col_colors, row_colors=row_colors,
> +                            figsize=(12, 12), method=method, metric=metric,
> +                            cbar_kws={'label': 'Conditional\nProbability'})
>
> I am adding this in the changes I will push up soon, but I am conflicted.
> Will the Conditionals always contain log conditional probabilities? Since
> currently the answer is yes, I have made this change, but we may want to
> consider if you foresee other types of data being contained in the
> Conditionals.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/biocore/rhapsody/pull/69?email_source=notifications&email_token=AA75VXOJXYDOLQN2QC5PHOLQCGFIHA5CNFSM4IIAXPL2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCAEJLTA#discussion_r309209841>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA75VXO5RABIW4CFUUQIRJDQCGFIHANCNFSM4IIAXPLQ>
> .
>

nbokulich commented:
Thanks @mortonjt ! I approved your changes and made a few changes in the latest commit:
1. I dropped score filtering per your request
2. I added metadata filtering per your request (i.e., if metadata files are provided only the intersection of IDs is shown)
3. I added a warning that is raised if metadata file(s) are input, indicating that the conditional probabilities will be filtered on the intersection of IDs.

mortonjt commented:
Thanks @nbokulich !
